src/components/Modals/Send/views/Address.tsx

```diff
@@ -72,6 +80,22 @@ export const Address = () => {
               validate: {
                 validateAddress: async (value: string) => {
                   const validAddress = await adapter.validateAddress(value)
+                  const validEnsAddress = await adapter.validateEnsAddress(value)
```

Comment:
@reidrankin What do you think about this? The `ChainAdapter` is unified across chain adapters so currently this fails type-checking but IMO, abstracting address validation and ENS addresses validation is cleaner and allows us not to repeat the same logic on the client (it's just a regex, but still).
I'm not sure or not whether it should be strictly unified, or if could support blockchain-specific additional methods like this one

Reply:
You should probably put it on `EthereumChainAdapter` and update the `ChainAdapter<T>` interface accordingly. (Something like `validateEnsAddress: (T extends ChainTypes.Ethereum ? (x: string) => boolean : unknown)`.) You will have to check that you're using the right type of adapter before you do the call, but that's a good thing. (I think you can just move this line down by one and solve that problem.)

Reply:
Sweet, thanks. That conditional is (almost) exactly what I needed for the typings.
That actually answers my concerns about `EthereumChainAdapter` having one more method than the others, but only partly, since other interface implementation won't pass type checks:
```shell
src/bitcoin/BitcoinChainAdapter.ts:44:14 - error TS2420: Class 'ChainAdapter' incorrectly implements interface 'ChainAdapter<ChainTypes.Bitcoin>'.
  Property 'validateEnsAddress' is missing in type 'ChainAdapter' but required in type 'ChainAdapter<ChainTypes.Bitcoin>'.
```

The cleanest I've found is to make `validateEnsAddress` optional and call it with the `?.` notation and access the result object members the same way. (`validEnsAddress?.valid`). That's not exactly true to the runtime however.

Reply:
Maybe `validateEnsAddress: (T extends ChainTypes.Ethereum ? (x: string) => boolean : undefined)` would do it -- though that's not going to be precisely future-proof given the advent of `exactOptionalPropertyTypes`. I'd suggest https://github.com/shapeshift/lib/compare/chain-specific-adapter-props-example

Reply:
> Sweet, thanks. That conditional is (almost) exactly what I needed for the typings. That actually answers my concerns about `EthereumChainAdapter` having one more method than the others, but only partly, since other interface implementation won't pass type checks:
> 
> ```shell
> src/bitcoin/BitcoinChainAdapter.ts:44:14 - error TS2420: Class 'ChainAdapter' incorrectly implements interface 'ChainAdapter<ChainTypes.Bitcoin>'.
>   Property 'validateEnsAddress' is missing in type 'ChainAdapter' but required in type 'ChainAdapter<ChainTypes.Bitcoin>'.
> ```
> 
> The cleanest I've found is to make `validateEnsAddress` optional and call it with the `?.` notation and access the result object members the same way. (`validEnsAddress?.valid`). That's not exactly true to the runtime however.

this is how we already do it for other methods like `signAndBroadcastTx` for metamask, i'm fine with this :)

Reply:
Looks like you can have your cake and eat it 🍰 https://github.com/shapeshift/lib/pull/315 + that's future-proof and not a regression on existing types. Unless there's an argument against it (readability ? using `Type` vs. `Interface`?) I believe that's a perfect solution